Sources/SwiftDocC/Infrastructure/DocumentationCurator.swift

```diff
@@ -229,10 +229,29 @@ struct DocumentationCurator {
                     (childDocumentationNode.kind == .article || childDocumentationNode.kind.isSymbol || childDocumentationNode.kind == .tutorial || childDocumentationNode.kind == .tutorialArticle) else {
                         continue
                 }
-                
-                guard childDocumentationNode.kind != .module else {
-                    problems.append(Problem(diagnostic: Diagnostic(source: source(), severity: .warning, range: range(), identifier: "org.swift.docc.ModuleCuration", summary: "Linking to \((link.destination ?? "").singleQuoted) from a Topics group in \(nodeReference.absoluteString.singleQuoted) isn't allowed", explanation: "The former is a module, and modules only exist at the root"), possibleSolutions: []))
+
+
+                // If a module has a path with a manual technology root there shouldn't
+                // be an error messsage. This is an if statement instead of a guard because
+                // when there's no warning we still curate the node
+                if childDocumentationNode.kind == .module {
+
+                    var hasTechnologyRoot = documentationNode.kind.isSymbol == false
```

Comment:
Since this isn't checking that the curating node is a module it is implicitly relying on the fact that articles aren't top-level pages which is a bit subtle. The two main ways to solve this would be

- Document the subtleties that this check relies on and describe why it produces the correct behavior
- Check the curating node the same as the first values in `context.pathsTo(...)`

Reply:
Could you please expand on what you mean by this? I'm still new to this codebase and how it works, but would love to learn more about it.

Reply:
Absolutely. If I haven't mentioned it already, when we say "curation" we may organizing documentation into a hierarchy by writing a link to another page in a "Topics sections". For example, this excerpt from the `@Metadata` documentation curates a handful of other pages

```md
## Topics

### Customizing the Presentation of a Page

- ``DisplayName``
- ``PageImage``
- ``PageKind``
...
```

The way that DocC processes the curation happens in 4 steps;

1. "crawling" the authored curation, starting with all top level symbols, nodes, and roots in order to relate nodes
2. adding all the remaining "uncurated" symbols and articles based on the symbol hierarchy (and articles are added top-level)
3. "crawling" any page that was added in step 2 since the crawler didn't reach that page in step 1.
4. checking if step 3 related any nodes that was also related in step 2 and prefer the authored curation (3) over the inferred "automatic" curation (2).

The way that both of those "crawl" step works its that the `DocumentationCurator` is given a list of references and for each reference it recursively descends the authored curation (the links to other pages in the authored topic sections) and relates the linked nodes with the page that has the curation.

---

Because step 1 only descends from top-level symbols, modules, and technology root pages; the only way that the curator would visit an article was if that article was curated somewhere. If the curator doesn't visit the article in step 1, the article gets added top-level to the only module page or root page, as long as there's not more than once. If this happens the curator will visit the article in step 3.
       
This means that if the curator is visiting an article, it either reached that article by descending from a module or technology root in step 1, or the article got "automatically" curated under a module/root page in step 2 and was visited explicitly in step 3.
       
Because of this, checking `var hasTechnologyRoot = documentationNode.kind.isSymbol == false` can only be true in two cases
- The node is a technology root
- The node is an article that has to have a module/root page as the first values of its `context.pathsTo(...)` (otherwise the curator wouldn't have reached the article)

So, in the article case, the check still does has the correct overall behavior but the reason why it's correct is very subtle.


One way to avoid this subtle but correct behavior would be to perform the same check on the current node and the first values in `context.pathsTo(...)`. This could have a structure that looks something like this:

```swift
func isTechnologyRoot(_ reference: ResolvedTopicReference) -> Bool {
    return // if the node for this reference is a technology root or not.
}

let hasTechnologyRoot = isTechnologyRoot(nodeReference) || context.pathsTo(nodeReference).contains(where: { path in
    guard let maybeRoot = path.first else { return false }
    return isTechnologyRoot(maybeRoot)
})
```

Reply:
Thank you for taking the time to explain the curating for me. After reading this more times than I'll like to admit 😅, and looking through some of the code base. I was able to wrap my brain around what's going on!!

If I'm understanding correctly, DocC curating is similar to a topological sort on a directed graph. Where the top level symbols lead to the recursive curation of all references they related too. Then this process repeats for any uncharted symbols/articles.